src/FSharpVSPowerTools.Logic/Outlining/OutliningTagger.fs

```diff
@@ -0,0 +1,196 @@
+
+            let missingLinesCount = Math.Max(nHintLines - MaxTooltipLines, 0)
+
+            let hintSnapshotSpan = SnapshotSpan(firstLine.Start, lastLine.End)
+            let hintText = lazy(
+                let text = hintSnapshotSpan.GetText()
+                match missingLinesCount with
+                | 0 -> text
+                | n -> text + sprintf "\n...\n\n +%d lines" n
+            )
+
+            TagSpan(
+                ss,
+                { new IOutliningRegionTag with
+                    member __.CollapsedForm      = "..." :> obj
+                    member __.CollapsedHintForm  = hintText.Force() :> obj
+                    member __.IsDefaultCollapsed = false
+                    member __.IsImplementation   = false })
+        with
+            | :? ArgumentOutOfRangeException -> null
+
+    let getTags (nssc: NormalizedSnapshotSpanCollection) =
+        if Seq.isEmpty nssc || Array.isEmpty snapshotSpans then Seq.empty
+        else
+            let newSnapshot = (Seq.head nssc).Snapshot
+            if newSnapshot.Version <> snapshotSpans.[0].Snapshot.Version then
+                snapshotSpans <- snapshotSpans |> Array.map (fun x -> x.TranslateTo(newSnapshot, SpanTrackingMode.EdgeExclusive))
+            snapshotSpans
+            |> Seq.filter nssc.IntersectsWith
+            |> Seq.map (createTagSpan)
+            |> Seq.cast<ITagSpan<_>>
+
+    let doUpdate () =
+        let uiContext = SynchronizationContext.Current
+        let worker =
+            asyncMaybe {
+                let dte = serviceProvider.GetService<EnvDTE.DTE, SDTE>()
+                let snapshot = buffer.CurrentSnapshot
+                let source = snapshot.GetText()
+                let! doc = dte.GetCurrentDocument(textDocument.FilePath)
+                let! project = projectFactory.CreateForDocument buffer doc
+                let! parseFileResults = languageService.ParseFileInProject (doc.FullName, source, project) |> Async.map Some
+                let! ast = parseFileResults.ParseTree
+                let ranges =
+                    ast
+                    |> visitAst 
+                    |> Seq.filter (fun r -> r.StartLine <> r.EndLine)
+                    |> Array.ofSeq
+                return (snapshot, ranges)
+            } |> Async.bind (fun x -> async {
+                 do! Async.SwitchToContext uiContext
+                 match x with
+                 | Some(snapshot, r) ->
+                    try
+                        let ss = r |> Array.map (fun x -> SnapshotSpan.OfRange(snapshot, x))
+                        triggerUpdate ss
+                    with
+                        | :? ArgumentOutOfRangeException -> ()
+                 | None -> ()
+            })
+
+        Async.StartInThreadPoolSafe worker
+
+    let docEventListener =
+        new DocumentEventListener(
+            [ViewChange.bufferEvent buffer],
+            UpdateDelay,
+            doUpdate)
+
+    do
+        doUpdate ()
```

Comment:
I believe we've discussed that this like is not needed since `DocumentEventListener` calls the callback immediately during its construction.